### DISCUSSION TITLE
PlatformIO.ini for BSB-LAN

### DISCUSSION THREAD

**Discussion by parsley** [fredlcore/BSB-LAN#671] [Q&A] [answered]
Hi,

because I use VScode / VScodium daily and kind of hate the "Arduino IDE", I reject installing it. 🙈😉 
I rather start tinkering until I find/create/fix a working platformio.ini file. This has not changed since I asked for it a few years ago [in issue 465](https://github.com/fredlcore/BSB-LAN/issues/465). Thankfully @fredlcore provided his version and I used it to install BSB-LAN Version 2.1.8-20220731102301 on my Due and never updated it since.
But things change and I'd like to switch to an Olimex ESP32-PoE-ISO powered by my network switch and also update BSB-LAN benefit from the new MQTT implementation.
In [issue 467](https://github.com/fredlcore/BSB-LAN/issues/467) also @henrythasler was asking for a platformio.ini but I do understand the concern @fredlcore explains in that issue, too. Maintaining a platformio.ini can be additional workload and that is why I now follow his suggestion and start this discussion. 
Right now I do not have any working platformio.ini and it would be greate if someone could post at least a fairly recent version that I could use as a starting point. The goal after that is obviously to keep this discussion alive and post a new and up to date platformio.ini version at least when someone asks for it.
@fredlcore could you help me out once again, please? 🫣

g n8, parsley

**Comment by fredlcore** [comment]
Sorry to say, but platformio for ESP32 is dead:
https://github.com/platformio/platform-espressif32/issues/1225
I hate that because I code with platformio/VSCode, but for normal users who don't/shouldn't tinker with the code (which will be overwritten with any new update anyways), installing Arduino IDE for the simple purpose of flashing the software, is the best option. You don't have to like it for that (and, btw., version 2.x is a great improvement compared to 1.8.x).

**Reply by parsley** [reply]
Hi

Yes, I recently learned about that issue and hate it since this kind of explains an issue I recently discovered with an ESP_NOW based pet project of mine. 
As far as I understand the issue PlatformIO did not update the espressif-arduino framework of PlatformIO for at least one year now. Which means all libraries are still on some version that is at least one year old and also newer ESP32 devices are not supported. That realy sucks.
I did not find any reason for this nor how to blame or how to fix or work around this mess since I didn't have the time to read all those ~400 Posts. Can someone summarize the issue?

However, you nevertheless still seam to develop BSB-LAN with PlatformIO, right?
If so, could you post your current setup despite the uncertain future of this setup?

g n8, parsley

**Comment by fredlcore** [comment, marked as answer]
I'm sorry, but you'll have look this up in the linked thread yourself. There are several ways of doing it, all are hacks, and I don't want to have to start entertaining questions about this here. It's really not needed for users of BSB-LAN, and if you want to continue using platformio for your own projects, then this understandable but off-topic here...

**Comment by parsley** [comment]
Hi

after reading all posts to the issue #1225 linked above (which took me way to long)... tl;dr: 😤💰💸😭🤬🤦‍♂️🤷‍♂️ (something between PlatformIO and Espressif about money and cooperation and some Kindergarden...)

...some (technically) rather interesting posts were:
https://github.com/platformio/platform-espressif32/issues/1225#issuecomment-2395047114
https://github.com/platformio/platform-espressif32/pull/1281#issuecomment-2468276598
https://github.com/espressif/arduino-esp32/discussions/10039
https://github.com/espressif/arduino-esp32

And at the end I have a platformio.ini file that seams to work just fine.
It is based on this one from @fredlcore: https://github.com/fredlcore/BSB-LAN/issues/465#issuecomment-1181740505

I wanted to flash a new Olimex ESP32-PoE-ISO board, so I copied the [env:esp32-EVB] section, changed board = esp32-poe-iso and platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.07/platform-espressif32.zip and had a working ini for this controller. 

For my old Due to compile I needed to add lib_deps = arduino-libraries/Ethernet @ ^2.0.1.

I don't need [env:esp32-EVB] and [env:esp32-NodeMCU] so I didn't fix or touch them at all but disabled them by setting default_envs = Due, esp32-Olimex-ESP-PoE-ISO. But if someone needs them they should be able to fix them theirself now. 😉

Here's my current file:

```
; PlatformIO Project Configuration File
;
;   Build options: build flags, source filter
;   Upload options: custom upload port, speed and extra flags
;   Library options: dependencies, extra library storages
;   Advanced options: extra scripting
;
; Please visit documentation for the other options and examples
; https://docs.platformio.org/page/projectconf.html

[platformio]
default_envs = Due, esp32-Olimex-ESP-PoE-ISO
src_dir = BSB_LAN

[env]
; common options
framework = arduino

[env:esp32-EVB]
;platform = espressif32
platform = https://github.com/platformio/platform-espressif32.git#feature/arduino-upstream
platform_packages = framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git
board = esp32-evb
build_src_filter = +<*> -<src/ArduinoMDNS> -<src/WiFiSpi>
upload_speed = 460800
board_build.partitions = min_spiffs.csv
build_flags = -DBSB_BUS
upload_port = /dev/cu.*usb*serial*
build_type = debug
monitor_filters = esp32_exception_decoder, time, log2file
monitor_speed = 115200

[env:Due]
platform = atmelsam
board = due
lib_deps =
  arduino-libraries/Ethernet @ ^2.0.1

[env:esp32-NodeMCU]
;platform = espressif32
platform = https://github.com/platformio/platform-espressif32.git#feature/arduino-upstream
platform_packages = framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git
board = esp32doit-devkit-v1
build_src_filter = +<*> -<src/ArduinoMDNS> -<src/WiFiSpi>
;upload_port = /dev/cu.SLAB_USBtoUART
upload_port = /dev/cu.usbserial*
;upload_speed = 921600
upload_speed = 115200
board_build.partitions = min_spiffs.csv
build_flags = -DNODEMCU

[env:esp32-Olimex-ESP-PoE-ISO]
; env made by LP
;platform = espressif32
;platform = https://github.com/platformio/platform-espressif32.git#feature/arduino-upstream
;platform_packages = framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git
;[env:stable]
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.07/platform-espressif32.zip
;[env:development]
;platform = https://github.com/pioarduino/platform-espressif32.git#develop
board = esp32-poe-iso
build_src_filter = +<*> -<src/ArduinoMDNS> -<src/WiFiSpi>
upload_speed = 460800
board_build.partitions = min_spiffs.csv
build_flags = -DBSB_BUS
upload_port = /dev/cu.*usb*serial*
build_type = debug
monitor_filters = esp32_exception_decoder, time, log2file
monitor_speed = 115200
```

**Comment by fredlcore** [comment]
Interesting. Does this fetch always the most recent version (3.0.7 IIRC) for you? I used the one that (I think) also wrote in one of the comments which contained the version number in the URL. It worked with 3.0.2, but when I updated this to a higher version, I got downloading errors, so I switched to the pioarduino repo, but since this is community-organized, I'm not sure if it will always work.
So, thanks for this, but as a note to anyone else reading this:

**I will not answer any questions that might arise from issues resulting from compiling BSB-LAN with platformio!**

I'm saddened by this as well because I was actually contemplating shifting the installation from Arduino IDE to platformio because many people don't read the manual and bug me with questions when they get a compile error because they didn't adjust the partition scheme, even though it is stated in three places in the manual. By just providing a platformio.ini for all platforms, this would be much easier, but given the uncertain state of affairs between platformio and espressif, I'm not going to redo all the videos and manuals for potentially nothing...

**Reply by parsley** [reply]
Sorry, it's complicated like you said. And I did not remove the inactive (commented) lines and the two [env:...] parts that I did not fix. So be carefull. ;)

The line that is active and works for me is

```platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.07/platform-espressif32.zip```

which is actually fixed to the latest release of pioarduino.
(Two lines under that is the platform link that claims to get the unstable/latest version. But I did not test it.)

My build output says:
```
PACKAGES: 
 - framework-arduinoespressif32 @ 3.0.7 
 - framework-arduinoespressif32-libs @ 5.1.0+sha.632e0c2a9f 
 - tool-esptoolpy @ 4.8.1 
 - tool-mklittlefs @ 3.2.0 
 - tool-riscv32-esp-elf-gdb @ 12.1.0+20221002 
 - tool-xtensa-esp-elf-gdb @ 12.1.0+20221002 
 - toolchain-xtensa-esp32 @ 12.2.0+20230208
 ```

Regarding "easy":
Nowadays big projects start to offer installing a precompiled binaries even by webinstallers.
For unexperienced users this would obviously be the easiest way. But this would mean to completely switch away form the concept of BSB_LAN_config.h and BSB_LAN_custom_defs.h and use individualised files on a SD instead.

**Reply by fredlcore** [reply]
Yes, for that reason, binaries are not an option as they would also prevent the custom code "plugins" that users have developed. I don't mind that users have to understand at least to some extent the configuration, but I would like to take away the need to understand how to configure the microcontrollers (which are manifold and also change over time without notice). For that, platformio.ini is a great way to do that as part of the project, but without a clear and reliable future path for platformio, I'm going to hold off on this for at least year or two to see which way it is going to be...
